File: mobile/lib/services/task_service.dart
import 'dart:convert';
import 'package:uuid/uuid.dart';
import '../models/task.dart';
import '../models/subtask.dart';
import '../models/team_member.dart';
import 'storage_service.dart';

class TaskService {
  // Récupérer toutes les tâches
  Future<List<Task>> getAllTasks() async {
    final tasksJson = StorageService.getAll('tasks');
    List<Task> tasks = [];

    for (var taskJson in tasksJson) {
      if (taskJson is String) {
        tasks.add(Task.fromJson(json.decode(taskJson)));
      }
    }

    return tasks;
  }

  // Récupérer les tâches d'un projet spécifique
  Future<List<Task>> getTasksByProject(String projectId) async {
    final allTasks = await getAllTasks();
    return allTasks.where((task) => task.projectId == projectId).toList();
  }

  // Récupérer une tâche spécifique
  Future<Task> getTaskById(String id) async {
    final taskJson = StorageService.get('tasks', id);
    if (taskJson == null) {
      throw Exception('Tâche non trouvée');
    }

    return Task.fromJson(json.decode(taskJson));
  }

  // Créer une nouvelle tâche
  Future<Task> createTask({
    required String projectId,
    required String title,
    required String description,
    TaskPriority priority = TaskPriority.medium,
    TeamMember? assignedTo,
    required String taskNumber,
    required String openedDate,
    required String openedBy,
  }) async {
    final id = const Uuid().v4();
    
    final task = Task(
      id: id,
      title: title,
      taskNumber: taskNumber,
      openedDate: openedDate,
      openedBy: openedBy,
      status: TaskStatus.open,
      timeSpent: '00:00:00',
      assignedTo: assignedTo,
      description: description,
      comments: 0,
      attachments: 0,
      projectId: projectId,
      kanbanStatus: 'backlog',
      priority: priority,
    );

    await StorageService.save('tasks', id, jsonEncode(task.toJson()));
    return task;
  }

  // Mettre à jour une tâche
  Future<Task> updateTask(Task task) async {
    await StorageService.save('tasks', task.id, jsonEncode(task.toJson()));
    return task;
  }

  // Mettre à jour le statut d'une tâche
  Future<Task> updateTaskStatus(String taskId, TaskStatus newStatus, Task updatedTask) async {
    await StorageService.save('tasks', taskId, jsonEncode(updatedTask.toJson()));
    return updatedTask;
  }

  // Supprimer une tâche
  Future<bool> deleteTask(String id) async {
    await StorageService.delete('tasks', id);
    return true;
  }

  // Ajouter une sous-tâche à une tâche
  Future<Task> addSubtask(String taskId, Subtask subtask) async {
    // Récupérer la tâche principale
    final taskJson = StorageService.get('tasks', taskId);
    if (taskJson == null) {
      throw Exception('Tâche non trouvée');
    }
    
    final task = Task.fromJson(json.decode(taskJson));
    
    // Ajouter la sous-tâche (on simule ceci, car notre modèle n'a pas de champ subtasks)
    // Dans une implémentation réelle, vous stockeriez les sous-tâches séparément
    
    // Pour l'exemple, nous allons stocker la sous-tâche dans une box dédiée
    await StorageService.save('subtasks', subtask.id, jsonEncode(subtask.toJson()));
    
    // Retourner la tâche mise à jour
    return task;
  }

  // Mettre à jour le statut d'une sous-tâche
  Future<Task> updateSubtaskStatus(
    String taskId,
    String subtaskId,
    SubtaskStatus status,
  ) async {
    // Récupérer la sous-tâche
    final subtaskJson = StorageService.get('subtasks', subtaskId);
    if (subtaskJson == null) {
      throw Exception('Sous-tâche non trouvée');
    }
    
    final subtask = Subtask.fromJson(json.decode(subtaskJson));
    
    // Créer une sous-tâche mise à jour
    final updatedSubtask = Subtask(
      id: subtask.id,
      title: subtask.title,
      description: subtask.description,
      status: status,
      priority: subtask.priority,
      timeSpent: subtask.timeSpent,
      assignedTo: subtask.assignedTo,
      createdAt: subtask.createdAt,
      dueDate: subtask.dueDate,
      comments: subtask.comments,
      attachments: subtask.attachments,
      parentTaskId: subtask.parentTaskId,
    );
    
    // Sauvegarder la sous-tâche mise à jour
    await StorageService.save('subtasks', subtaskId, jsonEncode(updatedSubtask.toJson()));
    
    // Récupérer la tâche principale
    final task = await getTaskById(taskId);
    
    return task;
  }
}

File: front/src/services/taskService.ts
import api from './api';
import { Task } from '../types';

interface CreateTaskData {
  title: string;
  description?: string;
  projectId: string;
  assignedTo?: string;
  status?: 'Open' | 'InProgress' | 'Completed' | 'Canceled';
  priority?: 'low' | 'medium' | 'high';
  dueDate?: string;
  kanbanStatus?: 'backlog' | 'in-progress' | 'completed';
}

interface UpdateTaskData {
  title?: string;
  description?: string;
  assignedTo?: string;
  status?: 'Open' | 'InProgress' | 'Completed' | 'Canceled';
  priority?: 'low' | 'medium' | 'high';
  dueDate?: string;
  kanbanStatus?: 'backlog' | 'in-progress' | 'completed';
}

class TaskService {
  // Récupérer toutes les tâches d'un projet
  async getTasksByProject(projectId: string): Promise<Task[]> {
    try {
      const response = await api.get<Task[]>(`/projects/${projectId}/tasks`);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la récupération des tâches du projet ${projectId}:`, error);
      throw error;
    }
  }

  // Récupérer une tâche par son ID
  async getTaskById(taskId: string): Promise<Task> {
    try {
      const response = await api.get<Task>(`/tasks/${taskId}`);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la récupération de la tâche ${taskId}:`, error);
      throw error;
    }
  }

  // Créer une nouvelle tâche
  async createTask(taskData: CreateTaskData): Promise<Task> {
    try {
      const response = await api.post<Task>('/tasks', taskData);
      return response.data;
    } catch (error) {
      console.error('Erreur lors de la création de la tâche:', error);
      throw error;
    }
  }

  // Mettre à jour une tâche
  async updateTask(taskId: string, taskData: UpdateTaskData): Promise<Task> {
    try {
      const response = await api.put<Task>(`/tasks/${taskId}`, taskData);
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la mise à jour de la tâche ${taskId}:`, error);
      throw error;
    }
  }

  // Supprimer une tâche
  async deleteTask(taskId: string): Promise<void> {
    try {
      await api.delete(`/tasks/${taskId}`);
    } catch (error) {
      console.error(`Erreur lors de la suppression de la tâche ${taskId}:`, error);
      throw error;
    }
  }

  // Changer le statut d'une tâche
  async updateTaskStatus(taskId: string, status: Task['status']): Promise<Task> {
    try {
      const response = await api.patch<Task>(`/tasks/${taskId}/status`, { status });
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de la mise à jour du statut de la tâche ${taskId}:`, error);
      throw error;
    }
  }

  // Assigner une tâche à un utilisateur
  async assignTask(taskId: string, userId: string): Promise<Task> {
    try {
      const response = await api.patch<Task>(`/tasks/${taskId}/assign`, { userId });
      return response.data;
    } catch (error) {
      console.error(`Erreur lors de l'assignation de la tâche ${taskId}:`, error);
      throw error;
    }
  }

  // Déplacer une tâche vers un statut kanban différent
  async moveTaskKanban(taskId: string, kanbanStatus: Task['kanbanStatus']): Promise<Task> {
    try {
      const response = await api.patch<Task>(`/tasks/${taskId}/kanban`, { kanbanStatus });
      return response.data;
    } catch (error) {
      console.error(`Erreur lors du déplacement de la tâche ${taskId} dans le kanban:`, error);
      throw error;
    }
  }
}

export default new TaskService();
